File: components/cell-mesh/src/cell_rules.c
// cell_rules.c — fixed-vocabulary rules engine. Pure C, no I/O.

#include "cell_rules.h"
#include "cell_wire.h"

#include <string.h>

// ── Wire format offsets (match the layout documented in cell_rules.h) ─

#define R_OFF_VERSION              0u
#define R_OFF_TRIGGER_KIND         1u
#define R_OFF_TRIGGER_TYPE_HASH    2u
#define R_OFF_QUORUM_N             34u
#define R_OFF_QUORUM_WINDOW_MS     35u
#define R_OFF_QUORUM_DISTINCT      37u
#define R_OFF_EFFECT_KIND          38u
#define R_OFF_BLINK_DURATION_MS    39u
#define R_OFF_EMIT_TYPE_HASH       41u
#define R_OFF_EMIT_PAYLOAD_LEN     73u
#define R_OFF_EMIT_PAYLOAD         75u

int cm_rule_encode(const cm_rule_t *rule, uint8_t out_buf[CM_RULE_ENCODED_SIZE]) {
    if (!rule || !out_buf) return -1;

    memset(out_buf, 0, CM_RULE_ENCODED_SIZE);
    out_buf[R_OFF_VERSION]      = CM_RULE_SCHEMA_VERSION;
    out_buf[R_OFF_TRIGGER_KIND] = (uint8_t)rule->trigger_kind;
    memcpy(out_buf + R_OFF_TRIGGER_TYPE_HASH, rule->trigger_type_hash, 32);

    if (rule->trigger_kind == CM_TRIGGER_QUORUM) {
        out_buf[R_OFF_QUORUM_N] = rule->quorum_n;
        cm_write_u16(out_buf + R_OFF_QUORUM_WINDOW_MS, rule->quorum_window_ms);
        out_buf[R_OFF_QUORUM_DISTINCT] = rule->quorum_distinct_peers ? 1 : 0;
    }

    out_buf[R_OFF_EFFECT_KIND] = (uint8_t)rule->effect.kind;
    switch (rule->effect.kind) {
        case CM_EFFECT_BLINK:
            cm_write_u16(out_buf + R_OFF_BLINK_DURATION_MS, rule->effect.as.blink.duration_ms);
            break;
        case CM_EFFECT_EMIT:
            memcpy(out_buf + R_OFF_EMIT_TYPE_HASH, rule->effect.as.emit.type_hash, 32);
            cm_write_u16(out_buf + R_OFF_EMIT_PAYLOAD_LEN, rule->effect.as.emit.payload_len);
            if (rule->effect.as.emit.payload_len > 0
                && rule->effect.as.emit.payload_len <= 64) {
                memcpy(out_buf + R_OFF_EMIT_PAYLOAD,
                       rule->effect.as.emit.payload,
                       rule->effect.as.emit.payload_len);
            }
            break;
        case CM_EFFECT_NONE:
        default:
            break;
    }
    return 0;
}

int cm_rule_decode(const uint8_t buf[CM_RULE_ENCODED_SIZE], cm_rule_t *out_rule) {
    if (!buf || !out_rule) return -1;
    if (buf[R_OFF_VERSION] != CM_RULE_SCHEMA_VERSION) return -1;

    memset(out_rule, 0, sizeof(*out_rule));
    out_rule->occupied = false; // caller installs to set this

    uint8_t tk = buf[R_OFF_TRIGGER_KIND];
    if (tk != CM_TRIGGER_ON_TYPE && tk != CM_TRIGGER_QUORUM) return -1;
    out_rule->trigger_kind = (cm_trigger_kind_t)tk;

    memcpy(out_rule->trigger_type_hash, buf + R_OFF_TRIGGER_TYPE_HASH, 32);

    if (tk == CM_TRIGGER_QUORUM) {
        out_rule->quorum_n              = buf[R_OFF_QUORUM_N];
        out_rule->quorum_window_ms      = cm_read_u16(buf + R_OFF_QUORUM_WINDOW_MS);
        out_rule->quorum_distinct_peers = buf[R_OFF_QUORUM_DISTINCT] != 0;
    }

    uint8_t ek = buf[R_OFF_EFFECT_KIND];
    if (ek != CM_EFFECT_BLINK && ek != CM_EFFECT_EMIT) return -1;
    out_rule->effect.kind = (cm_effect_kind_t)ek;

    switch (ek) {
        case CM_EFFECT_BLINK:
            out_rule->effect.as.blink.duration_ms = cm_read_u16(buf + R_OFF_BLINK_DURATION_MS);
            break;
        case CM_EFFECT_EMIT:
            memcpy(out_rule->effect.as.emit.type_hash, buf + R_OFF_EMIT_TYPE_HASH, 32);
            out_rule->effect.as.emit.payload_len = cm_read_u16(buf + R_OFF_EMIT_PAYLOAD_LEN);
            if (out_rule->effect.as.emit.payload_len > 64) return -1;
            memcpy(out_rule->effect.as.emit.payload,
                   buf + R_OFF_EMIT_PAYLOAD,
                   out_rule->effect.as.emit.payload_len);
            break;
        default:
            return -1;
    }
    return 0;
}

bool cm_rule_equals(const cm_rule_t *a, const cm_rule_t *b) {
    if (!a || !b) return false;
    uint8_t buf_a[CM_RULE_ENCODED_SIZE], buf_b[CM_RULE_ENCODED_SIZE];
    if (cm_rule_encode(a, buf_a) != 0) return false;
    if (cm_rule_encode(b, buf_b) != 0) return false;
    return memcmp(buf_a, buf_b, CM_RULE_ENCODED_SIZE) == 0;
}

void cm_rules_init(cm_rules_t *rules) {
    if (!rules) return;
    memset(rules, 0, sizeof(*rules));
}

int cm_rules_install(cm_rules_t *rules, const cm_rule_t *rule) {
    if (!rules || !rule) return -1;
    if (rule->trigger_kind == CM_TRIGGER_NONE) return -1;
    if (rule->effect.kind  == CM_EFFECT_NONE)  return -1;

    for (size_t i = 0; i < CM_RULES_MAX; i++) {
        if (!rules->entries[i].occupied) {
            rules->entries[i] = *rule;
            rules->entries[i].occupied = true;
            return (int)i;
        }
    }
    return -1;
}

int cm_rules_remove(cm_rules_t *rules, size_t slot) {
    if (!rules || slot >= CM_RULES_MAX) return -1;
    rules->entries[slot].occupied = false;
    return 0;
}

static bool trigger_matches(const cm_rule_t *r,
                            const cm_ring_t *ring,
                            const uint8_t *cell,
                            uint64_t now_ms) {
    switch (r->trigger_kind) {
        case CM_TRIGGER_ON_TYPE:
            return memcmp(cm_type_hash(cell), r->trigger_type_hash, 32) == 0;

        case CM_TRIGGER_QUORUM: {
            if (!ring) return false;
            // The just-arrived cell must already be in the ring; the
            // caller is responsible for pushing before evaluating.
            // First the type must match — saves a ring scan on irrelevant cells.
            if (memcmp(cm_type_hash(cell), r->trigger_type_hash, 32) != 0) return false;
            size_t n = cm_ring_count_recent(ring,
                                            r->trigger_type_hash,
                                            now_ms,
                                            r->quorum_window_ms,
                                            r->quorum_distinct_peers);
            return n >= r->quorum_n;
        }

        case CM_TRIGGER_NONE:
        default:
            return false;
    }
}

size_t cm_rules_evaluate(cm_rules_t *rules,
                          const cm_ring_t *ring,
                          const uint8_t cell[CM_CELL_SIZE],
                          uint64_t now_ms,
                          cm_effect_t out_effects[CM_RULES_MAX]) {
    if (!rules || !cell || !out_effects) return 0;

    rules->total_evaluated++;

    size_t count = 0;
    for (size_t i = 0; i < CM_RULES_MAX; i++) {
        const cm_rule_t *r = &rules->entries[i];
        if (!r->occupied) continue;
        if (!trigger_matches(r, ring, cell, now_ms)) continue;
        out_effects[count++] = r->effect;
        rules->total_fired++;
    }
    return count;
}

File: components/cell-mesh-zig/src/cell_rules.zig
const std = @import("std");
const wire = @import("cell_wire.zig");
const ring_mod = @import("cell_ring.zig");

pub const rules_max: usize = 8;
pub const encoded_size: usize = 139;
pub const schema_version: u8 = 0x01;

const trigger_none: c_int = 0;
const trigger_on_type: c_int = 1;
const trigger_quorum: c_int = 2;

const effect_none: c_int = 0;
const effect_blink: c_int = 1;
const effect_emit: c_int = 2;

pub const BlinkEffect = extern struct {
    duration_ms: u16,
};

pub const EmitEffect = extern struct {
    type_hash: [32]u8,
    payload_len: u16,
    payload: [64]u8,
};

pub const EffectAs = extern union {
    blink: BlinkEffect,
    emit: EmitEffect,
};

pub const Effect = extern struct {
    kind: c_int,
    as: EffectAs,
};

pub const Rule = extern struct {
    occupied: bool,
    trigger_kind: c_int,
    trigger_type_hash: [32]u8,
    quorum_n: u8,
    quorum_window_ms: u16,
    quorum_distinct_peers: bool,
    effect: Effect,
};

pub const Rules = extern struct {
    entries: [rules_max]Rule,
    total_evaluated: u32,
    total_fired: u32,
};

pub export fn cm_rule_encode(maybe_rule: ?*const Rule, maybe_out_buf: ?[*]u8) callconv(.c) c_int {
    const rule = maybe_rule orelse return -1;
    const out_buf = maybe_out_buf orelse return -1;

    @memset(out_buf[0..encoded_size], 0);
    out_buf[0] = schema_version;
    out_buf[1] = @intCast(rule.trigger_kind & 0xff);
    @memcpy(out_buf[2..34], rule.trigger_type_hash[0..]);

    if (rule.trigger_kind == trigger_quorum) {
        out_buf[34] = rule.quorum_n;
        wire.writeU16(out_buf[35..][0..2], rule.quorum_window_ms);
        out_buf[37] = if (rule.quorum_distinct_peers) 1 else 0;
    }

    out_buf[38] = @intCast(rule.effect.kind & 0xff);
    switch (rule.effect.kind) {
        effect_blink => wire.writeU16(out_buf[39..][0..2], rule.effect.as.blink.duration_ms),
        effect_emit => {
            @memcpy(out_buf[41..73], rule.effect.as.emit.type_hash[0..]);
            wire.writeU16(out_buf[73..][0..2], rule.effect.as.emit.payload_len);
            if (rule.effect.as.emit.payload_len > 0 and rule.effect.as.emit.payload_len <= 64) {
                const n: usize = rule.effect.as.emit.payload_len;
                @memcpy(out_buf[75..][0..n], rule.effect.as.emit.payload[0..n]);
            }
        },
        else => {},
    }
    return 0;
}

pub export fn cm_rule_decode(maybe_buf: ?[*]const u8, maybe_out_rule: ?*Rule) callconv(.c) c_int {
    const buf = maybe_buf orelse return -1;
    const out_rule = maybe_out_rule orelse return -1;
    if (buf[0] != schema_version) return -1;

    out_rule.* = std.mem.zeroes(Rule);
    const tk = buf[1];
    if (tk != trigger_on_type and tk != trigger_quorum) return -1;
    out_rule.trigger_kind = tk;
    @memcpy(out_rule.trigger_type_hash[0..], buf[2..34]);

    if (tk == trigger_quorum) {
        out_rule.quorum_n = buf[34];
        out_rule.quorum_window_ms = wire.readU16(buf[35..][0..2]);
        out_rule.quorum_distinct_peers = buf[37] != 0;
    }

    const ek = buf[38];
    if (ek != effect_blink and ek != effect_emit) return -1;
    out_rule.effect.kind = ek;
    switch (ek) {
        effect_blink => out_rule.effect.as.blink.duration_ms = wire.readU16(buf[39..][0..2]),
        effect_emit => {
            @memcpy(out_rule.effect.as.emit.type_hash[0..], buf[41..73]);
            out_rule.effect.as.emit.payload_len = wire.readU16(buf[73..][0..2]);
            if (out_rule.effect.as.emit.payload_len > 64) return -1;
            const n: usize = out_rule.effect.as.emit.payload_len;
            @memcpy(out_rule.effect.as.emit.payload[0..n], buf[75..][0..n]);
        },
        else => return -1,
    }
    return 0;
}

pub export fn cm_rule_equals(maybe_a: ?*const Rule, maybe_b: ?*const Rule) callconv(.c) bool {
    const a = maybe_a orelse return false;
    const b = maybe_b orelse return false;
    var buf_a: [encoded_size]u8 = undefined;
    var buf_b: [encoded_size]u8 = undefined;
    if (cm_rule_encode(a, &buf_a) != 0) return false;
    if (cm_rule_encode(b, &buf_b) != 0) return false;
    return std.mem.eql(u8, buf_a[0..], buf_b[0..]);
}

pub export fn cm_rules_init(maybe_rules: ?*Rules) callconv(.c) void {
    const rules = maybe_rules orelse return;
    rules.* = std.mem.zeroes(Rules);
}

pub export fn cm_rules_install(maybe_rules: ?*Rules, maybe_rule: ?*const Rule) callconv(.c) c_int {
    const rules = maybe_rules orelse return -1;
    const rule = maybe_rule orelse return -1;
    if (rule.trigger_kind == trigger_none) return -1;
    if (rule.effect.kind == effect_none) return -1;

    for (&rules.entries, 0..) |*entry, i| {
        if (!entry.occupied) {
            entry.* = rule.*;
            entry.occupied = true;
            return @intCast(i);
        }
    }
    return -1;
}

pub export fn cm_rules_remove(maybe_rules: ?*Rules, slot: usize) callconv(.c) c_int {
    const rules = maybe_rules orelse return -1;
    if (slot >= rules_max) return -1;
    rules.entries[slot].occupied = false;
    return 0;
}

fn triggerMatches(r: *const Rule, maybe_ring: ?*const ring_mod.Ring, cell: [*]const u8, now_ms: u64) bool {
    switch (r.trigger_kind) {
        trigger_on_type => return std.mem.eql(u8, cell[wire.Off.type_hash..][0..32], r.trigger_type_hash[0..]),
        trigger_quorum => {
            const ring = maybe_ring orelse return false;
            if (!std.mem.eql(u8, cell[wire.Off.type_hash..][0..32], r.trigger_type_hash[0..])) return false;
            const n = ring_mod.cm_ring_count_recent(
                ring,
                @ptrCast(&r.trigger_type_hash),
                now_ms,
                r.quorum_window_ms,
                r.quorum_distinct_peers,
            );
            return n >= r.quorum_n;
        },
        else => return false,
    }
}

pub export fn cm_rules_evaluate(
    maybe_rules: ?*Rules,
    maybe_ring: ?*const ring_mod.Ring,
    maybe_cell: ?[*]const u8,
    now_ms: u64,
    maybe_out_effects: ?[*]Effect,
) callconv(.c) usize {
    const rules = maybe_rules orelse return 0;
    const cell = maybe_cell orelse return 0;
    const out_effects = maybe_out_effects orelse return 0;

    rules.total_evaluated +%= 1;

    var count: usize = 0;
    for (&rules.entries) |*r| {
        if (!r.occupied) continue;
        if (!triggerMatches(r, maybe_ring, cell, now_ms)) continue;
        out_effects[count] = r.effect;
        count += 1;
        rules.total_fired +%= 1;
    }
    return count;
}
